### ISSUE TITLE
Unfriendly handling of patterns containing match with holes

### ISSUE THREAD

**Issue by silene** [coq/coq#6867]
#### Version

8.7

#### Description of the problem

As a user, I feel that the way Coq handles patterns containing `match` constructs with holes is needlessly painful. Here is an example:

```coq
Lemma foo x : x > match x with O => O | S n => n end.
replace (match x with _ => _ end) with (pred x).
```

Expected goal: `x > pred x`

Actual goal: `pred x > match pred x with O => O | S n => n end`

Instead of replacing the `match` subterm of the goal, Coq has replaced the first subterm that has the same type as `pred x`, i.e. `x`. In other words, Coq has completely ignored the pattern given by the user. Here is a variant:

```coq
Lemma foo x : x > match x with O => O | S n => n end.
set (y := match x with _ => _ end).
```

Expected goal: `x > y`.

Actual goal: `y`.

In both cases, the issue can be worked around by putting more information in the pattern to prevent Coq from ignoring it, e.g. `match x with O => _ | _ => _ end` (notice the `O` branch). Presumably, this issue is related to the way Coq handles the base case when internalizing nested `match` constructs.

**Comment by ppedrot** [comment]
Maybe we would need a syntax to denote a pattern-matching with unspecified branches? Like e.g. `match M with _ end`. Otherwise I don't see how you can get what you want without ambiguity.

**Comment by Zimmi48** [comment]
@ppedrot That's something that I had in mind as well. But that would require changing `constr`.

**Comment by ppedrot** [comment]
@Zimmi48 I don't think so. Such tactics actually take a pattern as argument (maybe using the infamous `pattern_of_constr`), so changing `glob_constr` should be enough.

**Comment by Zimmi48** [comment]
OK it depends what the end goal is. If a tactic needs to be able to leave a partial proof-term with a hole denoting unspecified branches like I had in mind, then you have to.

**Comment by gares** [comment]
@maximedenes once explained me that the term `match x with y => y end` is "elaborated" to "x", maybe this is just an occurrence of that?  If it was the case, then the tactic would just be called as "replace x with pred x" that is what it seems to do.

**Comment by silene** [comment]
> maybe this is just an occurrence of that?

No, that is not what happens. Note that the pattern is `match x with _ => _ end`, which Coq happily elaborates to `_`. This is easily seen on the second example which uses `set` instead of `replace`.

**Comment by mattam82** [comment]
It reminds me of another unrelated thing. @cmangin I can't find it again but maybe it was in your development that we found out that this feature of Coq's elaborator (match x with Foo _ => bar | y => p y end) changes the default branch to (p x) broke when trying to build a fixpoint consicely (repeating the map_constr_with_binders OCaml/Coq idiom). We should make a separate issue about that with the example
